pyiceberg/table/upsert_util.py

```diff
@@ -39,7 +77,7 @@ def create_match_filter(df: pyarrow_table, join_cols: list[str]) -> BooleanExpre
             functools.reduce(operator.and_, [EqualTo(col, row[col]) for col in join_cols]) for row in unique_keys.to_pylist()
         ]
 
-        return AlwaysFalse() if len(filters) == 0 else functools.reduce(operator.or_, filters)
+        return AlwaysFalse() if len(filters) == 0 else build_balanced_tree(operator.or_, filters)
```

Comment:
I'm thinking out loud to maybe bring this logic into the `Or` operation:

https://github.com/apache/iceberg-python/blob/71cb2473bee8161372b057207c82f73d42ae3092/pyiceberg/expressions/__init__.py#L260

We could pass it in like:
```suggestion
        return AlwaysFalse() if len(filters) == 0 else Or(*filters)
```

By bringing this into the `Op` operation, then this is much more structured. Thoughts/opinions?

Reply:
Yes! That's a great change. And this change will only affect when you pass more than 2 expressions, so you can still keep full control if you want to build an unbalanced tree yourself.

Let me change this.